pkg/controller/runGC.go

```diff
@@ -0,0 +1,154 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8s "k8s.io/client-go/kubernetes"
+	ctrlzap "sigs.k8s.io/controller-runtime/pkg/log/zap"
+)
+
+type ImageBuildGC struct {
+	maxIBRetention int
+	hephClient     clientset.Interface
+	namespaces     []string
+}
+
+func NewImageBuildGC(maxIBRetention int, log logr.Logger, ibNamespaces []string) (*ImageBuildGC, error) {
+	config, err := kubernetes.RestConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	hephClient, _ := clientset.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	var ns []string
+	if len(ibNamespaces) > 0 {
+		log.Info("Image Build GC limiting namespaces", "namespaces", ibNamespaces)
+		ns = ibNamespaces
+	} else {
+		// create k8s client to access all namespaces in the cluster
+		k8sClient, err := k8s.NewForConfig(config)
+		if err != nil {
+			return nil, err
+		}
+		ns, err = getAllNamespaces(k8sClient)
+		if err != nil {
+			log.Info("Unable to access cluster namespaces")
+			return nil, err
+		}
+		log.Info("Running GC against all namespaces")
+	}
+
+	return &ImageBuildGC{
+		maxIBRetention: maxIBRetention,
+		hephClient:     hephClient,
+		namespaces:     ns,
+	}, nil
+}
+
+func getAllNamespaces(k8sClient k8s.Interface) ([]string, error) {
+	namespaces, err := k8sClient.CoreV1().Namespaces().List(context.Background(), metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+	var ns []string
+	for _, namespace := range namespaces.Items {
+		ns = append(ns, namespace.Name)
+	}
+	return ns, nil
+}
+
+func (gc *ImageBuildGC) GCImageBuilds(ctx context.Context, log logr.Logger, namespace string) error {
+	crdList, err := gc.hephClient.HephaestusV1().ImageBuilds(namespace).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		log.Info("Unable to access a list of IBs, not starting Image Build GC", "error", err)
+		return err
+	}
+
+	listLen := len(crdList.Items)
+	if listLen == 0 {
+		log.V(1).Info("No build resources found, aborting")
+		return nil
+	}
+
+	var builds []hephv1.ImageBuild
+	for _, ib := range crdList.Items {
+		state := ib.Status.Phase
+		if state == hephv1.PhaseFailed || state == hephv1.PhaseSucceeded {
+			builds = append(builds, ib)
+		}
+	}
+
+	if len(builds) <= gc.maxIBRetention {
+		log.Info("Total resources are less than or equal to retention limit, aborting",
+			"resourceCount", len(builds), "retentionCount", gc.maxIBRetention)
+		return nil
+	}
+	log.Info("Total resources eligible for deletion", "count", len(builds))
+	sort.Slice(builds, func(i, j int) bool {
+		return builds[i].CreationTimestamp.Before(&builds[j].CreationTimestamp)
+	})
+
+	var errList []error
+	deletePolicy := metav1.DeletePropagationForeground
+	for _, build := range builds[:len(builds)-gc.maxIBRetention] {
+		if err := gc.hephClient.HephaestusV1().ImageBuilds(namespace).Delete(ctx, build.Name,
+			metav1.DeleteOptions{PropagationPolicy: &deletePolicy}); err != nil {
+			log.Info("Failed to delete build", "name", build.Name, "namespace", build.Namespace, "error", err)
```

Comment:
Is this log line redundant with the returned error list that's logged when the program exits?